### ISSUE TITLE
ULD loss ignores the response-only label mask and causal shift

### ISSUE THREAD

**Issue by Amix29** [MakazhanAlpamys/Soup#682]
## Description

The SFT cross-entropy respects causal shifting and `labels == -100`, but the ULD term receives
only `attention_mask`. It therefore optimizes prompt or other positions explicitly excluded by
Soup's supervision mask, and also includes positions outside the shifted causal targets.

Affected revision: `ef2e02b51a1b34546823edce55715d28234805cb` (Soup 0.74.0).

Relevant code:

- [cross-entropy uses shifted labels and `ignore_index=-100`](https://github.com/MakazhanAlpamys/Soup/blob/ef2e02b51a1b34546823edce55715d28234805cb/src/soup_cli/trainer/distill.py#L505-L517)
- [ULD receives only `attention_mask`](https://github.com/MakazhanAlpamys/Soup/blob/ef2e02b51a1b34546823edce55715d28234805cb/src/soup_cli/trainer/distill.py#L635-L642)
- [`uld_distill_loss` masks only padding/attention positions](https://github.com/MakazhanAlpamys/Soup/blob/ef2e02b51a1b34546823edce55715d28234805cb/src/soup_cli/utils/uld.py#L228-L281)

## Steps to reproduce

Run the original loss functions with real FP32 tensors and:

```text
labels = [-100, -100, 2]
```

The logit at position 0 predicts a target excluded by the causal label mask.

```text
gradient L1 at position 0, CE-only control: 0.0
gradient L1 at position 0, current CE + ULD: ~0.075376
```

## Expected behavior

When `train_on_responses_only` or another label mask excludes a target, every configured loss
term should follow the same causal supervision boundary unless a separate behavior is explicitly
configured and documented.

## Impact

The optimized objective differs from the requested one. Prompt tokens and other excluded spans
can influence adapter updates even though the SFT term correctly ignores them.

## Proposed fix and acceptance criteria

Build an explicit causal supervision mask from shifted labels and propagate it through ULD,
including the aligned-tokenizer mapping.

- [ ] Prompt-only positions contribute zero ULD gradient in responses-only mode.
- [ ] Padding, multi-turn conversations, and mixed masked/unmasked batches are covered.
- [ ] Entirely masked batches remain finite and produce no spurious update.
- [ ] The aligned ULD strategy maps the supervision mask through token alignment.

## Environment

Reproduced under Python 3.12.2 and PyTorch 2.12.1 CPU with the original Soup loss code and
autograd. No full CUDA training or final-quality comparison was performed.

**Comment by MakazhanAlpamys** [comment]
**Confirmed, both halves, and they are separable.**

Executed against the real trainer, gradient captured on `student_logits`:

```
labels            : [-100,-100,-100,-100,-100,-100, 3, 7, 4, 10, 5]
position          : [   0,   1,   2,   3,   4,   5, 6, 7, 8,  9,10]
supervised by CE? : [ n, n, n, n, n, Y, Y, Y, Y, Y, n(last) ]

|grad| CE only    : [0.0, 0.0, 0.0, 0.0, 0.0, 0.366, ... , 0.0]
|grad| CE + ULD   : [0.136, 0.136, 0.138, 0.136, 0.136, 0.238, ... , 0.136]

sum |grad| over CE-masked positions, CE only : 0.000000
sum |grad| over CE-masked positions, CE+ULD  : 0.817366
```

Positions 0-4 are **prompt** tokens the label mask excluded; position 10 is the **shift** artefact, since CE drops the final logit and ULD keeps it. Same result for all three strategies — `wasserstein`, `topk_align` and `wasserstein_aligned` all give the identical masked-position sum of 0.817366, and `uld_aligned_loss` has no `labels` parameter in its signature either.

## The strongest evidence is internal inconsistency, and it points at the fix

In the *same* `compute_loss`, the other two distillation terms do get labels:

```python
_compute_distill_term(..., labels=labels, attention_mask=...)
minillm_cb.distill_term(student_logits, teacher_logits, labels)
```

And `_compute_distill_term`'s own docstring states the reasoning:

> restricted to the trained tokens: `labels != -100` when labels is given (excludes padding AND prompt) ... Averaging over padding/prompt tokens (the pre-fix behaviour) diluted the signal.

So this exact fix was made deliberately on the plain-KL path and never propagated to the ULD branch, which passes `attention_mask` alone. That is a much better starting point for a fixer than a blank page: the intended semantics are already written down and already implemented next door.

## Severity

`data.train_on_responses_only` defaults to **True**, so this is the default configuration for `task: distill` with any ULD strategy, not an opt-in. Silent — the loss is finite and the curve looks normal.

Labelled `bug` + `help wanted`, unclaimed.

Related: #681, which I also confirmed today — the non-aligned strategies additionally feed the teacher student token ids, clamped. Different root cause, same two strategies, and a fixer touching `compute_loss` for either will be in the same lines. Worth reading both before starting.

Not settled: I established the gradient consequence, not a training-quality one. Nobody has shown a degraded final model, and you did not claim otherwise.

**Comment by AmirF194** [comment]
Taking this one.

**Comment by MakazhanAlpamys** [comment]
Linking the pull request onto the issue, which I should have done when it appeared: **#711** by @AmirF194.

Status: the fix mirrors the shift-and-mask block already shipped in `_compute_distill_term`, whose own docstring describes this exact bug class for the plain-KL path — the ULD branch was simply missed when that was fixed. Both halves of your report are closed and each is pinned separately.

The mutation I most expected to survive does not: **inverting the causal shift** (`labels[:, 1:]` -> `labels[:, :-1]`) fails 3 tests. A shift test is worthless unless its fixture makes shifted and unshifted give different numbers, and `labels=[[-100, 5, 2]]` is discriminating by construction. Verified numerically too, against an independently recomputed W1: `0.94520140` both ways. And the property the issue is actually about — per-position gradient L1 at the prompt position goes from `0.343096` on `main` to **exactly `0.0`**.

One detail worth recording here because it would be invisible otherwise: in `uld_aligned_loss` the teacher logits are deliberately *not* shifted, since they live in teacher token space and are aggregated onto student positions by `align`, while in `uld_distill_loss` both are shifted because they share a token space. Getting that backwards produces a plausible number and a wrong gradient.

**Comment by MakazhanAlpamys** [comment]
Fixed in `c5c0f32` by @AmirF194 in #711, landed after #704.

Both halves are closed and each is pinned separately. The property this issue is about, measured:

```
main:  per-position grad L1 = [0.343096, 0.442297, 0.406747]   prompt pos 0 = 0.343096
fixed: per-position grad L1 = [0.0, 0.587038, 0.653393, 0.0]   prompt pos 0 = 0.0 exactly
```

Five mutations killed. **The one worth recording is the off-by-one** — `labels[:, 1:]` to `labels[:, :-1]` — which fails 3 tests. A causal-shift test proves nothing unless its fixture makes shifted and unshifted give different numbers, and most fixtures do not; `labels=[[-100, 5, 2]]` is discriminating by construction, because unshifted masking leaves the final position supervised where shifted masking drops it. Verified numerically too, against an independently recomputed W1: `0.94520140` both ways.

**One asymmetry recorded here because it would otherwise be invisible to the next reader:** in `uld_aligned_loss` the teacher logits are deliberately **not** shifted, since they live in teacher token space and are aggregated onto student positions by `align`, while in `uld_distill_loss` both are shifted because they share a token space. Getting that backwards produces a plausible number and a wrong gradient.

Two acceptance criteria are not pinned by tests — mixed masked/unmasked batches, and multi-turn; every test uses `batch=1`. I checked both by hand and they behave correctly (gradient lands only on supervised, shifted positions; a fully-masked row gets exactly zero), so that is a test gap rather than a defect, and it is stated rather than glossed.

Reported by @Amix29.
